src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestControlled1.ll

```diff
@@ -27,8 +27,13 @@ entry:
   %20 = call %Callable* @__quantum__rt__callable_copy(%Callable* %19)
   call void @__quantum__rt__callable_make_controlled(%Callable* %20)
   call void @__quantum__rt__callable_invoke(%Callable* %20, %TupleHeader* %13, %TupleHeader* %result-tuple)
-  call void @__quantum__rt__tuple_unreference(%TupleHeader* %5)
-  call void @__quantum__rt__tuple_unreference(%TupleHeader* %13)
+  %21 = bitcast { %TupleHeader, %Qubit*, i64 }* %6 to %TupleHeader*
```

Comment:
I've commented inline the problems I'm seeing with this method. BTW, would it be possible for QIR generation to inject comments that tie the blocks of generated QIR to the source Q#?
```
define void @Lifted__PartialApplication__2__ctl__wrapper(%TupleHeader* %capture-tuple, %TupleHeader* %arg-tuple, %TupleHeader* %result-tuple) {
entry:
  ; extract inputs into typed variables
  %0 = bitcast %TupleHeader* %capture-tuple to { %TupleHeader, %Callable* }*
  %1 = bitcast %TupleHeader* %arg-tuple to { %TupleHeader, %Array*, %TupleHeader* }*
  %2 = getelementptr { %TupleHeader, %Array*, %TupleHeader* }, { %TupleHeader, %Array*, %TupleHeader* }* %1, i64 0, i32 1 ; array of controls from arg-tuple
  %3 = getelementptr { %TupleHeader, %Array*, %TupleHeader* }, { %TupleHeader, %Array*, %TupleHeader* }* %1, i64 0, i32 2 ; tupleheader from arg-tuple
  %4 = bitcast %TupleHeader** %3 to { %TupleHeader, %Qubit*, i64 }* ; inner tuple from arg-tuple

  ; ======
  ; replicate the inner tuple from arg-tuple
  %5 = call %TupleHeader* @__quantum__rt__tuple_create(
      i64 ptrtoint ({ %TupleHeader, %Qubit*, i64 }* 
      getelementptr ({ %TupleHeader, %Qubit*, i64 }, { %TupleHeader, %Qubit*, i64 }* null, i32 1) to i64))

  ; copy qubit* from inner arg-tuple into the new tuple
  %6 = bitcast %TupleHeader* %5 to { %TupleHeader, %Qubit*, i64 }*
  %7 = getelementptr { %TupleHeader, %Qubit*, i64 }, { %TupleHeader, %Qubit*, i64 }* %6, i64 0, i32 1 ; qubit* in new tuple
  %8 = getelementptr { %TupleHeader, %Qubit*, i64 }, { %TupleHeader, %Qubit*, i64 }* %4, i64 0, i32 1 ; qubit* in inner arg-tuple
  %9 = load %Qubit*, %Qubit** %8
  store %Qubit* %9, %Qubit** %7

  ; copy i64 from inner arg-tuple into the new tuple
  %10 = getelementptr { %TupleHeader, %Qubit*, i64 }, { %TupleHeader, %Qubit*, i64 }* %6, i64 0, i32 2
  %11 = getelementptr { %TupleHeader, %Qubit*, i64 }, { %TupleHeader, %Qubit*, i64 }* %4, i64 0, i32 2
  %12 = load i64, i64* %11
  store i64 %12, i64* %10

  ; ======
  ; replicate the arg-tuple (notice, that it doesn't addref the array of controls from arg-tuple)
  %13 = call %TupleHeader* @__quantum__rt__tuple_create(
      i64 ptrtoint ({ %TupleHeader, %Array*, %TupleHeader* }* 
      getelementptr ({ %TupleHeader, %Array*, %TupleHeader* }, { %TupleHeader, %Array*, %TupleHeader* }* null, i32 1) to i64))

  %14 = bitcast %TupleHeader* %13 to { %TupleHeader, %Array*, %TupleHeader* }*
  %15 = getelementptr { %TupleHeader, %Array*, %TupleHeader* }, { %TupleHeader, %Array*, %TupleHeader* }* %14, i64 0, i32 1
  %16 = load %Array*, %Array** %2
  store %Array* %16, %Array** %15
  %17 = getelementptr { %TupleHeader, %Array*, %TupleHeader* }, { %TupleHeader, %Array*, %TupleHeader* }* %14, i64 0, i32 2
  store %TupleHeader* %5, %TupleHeader** %17

  ; ======
  ; create a controlled callable from the capture-tuple and invoke it with the replicated arg-tuple
  ; (why do we need to replicate the arg-tuple at all for this?)
  %18 = getelementptr { %TupleHeader, %Callable* }, { %TupleHeader, %Callable* }* %0, i64 0, i32 1 ; callable** from capture-tuple
  %19 = load %Callable*, %Callable** %18  ; callable* from capture-tuple
  %20 = call %Callable* @__quantum__rt__callable_copy(%Callable* %19) ; new callable
  call void @__quantum__rt__callable_make_controlled(%Callable* %20)
  call void @__quantum__rt__callable_invoke(%Callable* %20, %TupleHeader* %13, %TupleHeader* %result-tuple)

  ; ======
  ; cleanup
  call void @__quantum__rt__tuple_unreference(%TupleHeader* %5) ; ok, release the copy of inner tuple
  call void @__quantum__rt__tuple_unreference(%TupleHeader* %13) ; ok, release the copy of arg-tuple
  %21 = bitcast { %TupleHeader, %Qubit*, i64 }* %6 to %TupleHeader* ; %21 == %5
  call void @__quantum__rt__tuple_unreference(%TupleHeader* %21) ; not ok! double release of %5
  %22 = bitcast { %TupleHeader, %Array*, %TupleHeader* }* %14 to %TupleHeader* ; %22 == %13
  call void @__quantum__rt__tuple_unreference(%TupleHeader* %22) ; not ok! double release of %13
  %23 = getelementptr { %TupleHeader, %Array*, %TupleHeader* }, { %TupleHeader, %Array*, %TupleHeader* }* 
    %14, i64 0, i32 1 ; not ok! %14 has been already released (twice at that)
  %24 = load %Array*, %Array** %23
  ; more likely than not, not ok! the array was never addref'ed by this method, thus it might
  ; release it but the caller might be still using it (even if you move this code above the release of %14)
  call void @__quantum__rt__array_unreference(%Array* %24)

  call void @__quantum__rt__callable_unreference(%Callable* %20) ; ok
  ret void
}
```

Reply:
I believe this should now be addressed and cleaned-up.